protocol/xmrtaker/net.go

```diff
@@ -37,6 +37,15 @@ func (inst *Instance) InitiateProtocol(
 	offer *types.Offer,
 ) (common.SwapState, error) {
 	expectedAmount, err := offer.ExchangeRate.ToXMR(providesAmount)
+
+	if expectedAmount.Cmp(offer.MinAmount) < 0 {
+		return nil, errAmountProvidedTooLow{providesAmount, offer.MinAmount}
+	}
+
+	if expectedAmount.Cmp(offer.MaxAmount) > 0 {
+		return nil, errAmountProvidedTooHigh{providesAmount, offer.MaxAmount}
+	}
+
```

Comment:
It would be good to call `coins.ValidatePositive` before doing the other checks. [This PR](https://github.com/AthanorLabs/atomic-swap/pull/380) adds an additional check to that method. Not taken care of in that PR, but scheduled for follow-up work, is handling max decimal places for ERC20 tokens. We limit the decimal places of the exchange rates to 6, and Monero is capped at 12 decimals. That means we can't exceed Ethereum's 18 decimal places, so forward and backwards exchange rate calculations will always equal each other.

We need some similar guarantee for tokens, or come up with a different system that doesn't require calculating the exchange rate in both directions. My thought for the follow-up work was to add code that limits the combined number of decimals used on the min/max values + the number of decimals on the exchange rate, and not allowing the sum to exceed the supported number of decimals on the ERC20 token. If you have any better ideas, let me know.

Reply:
> It would be good to call `coins.ValidatePositive`

Nice catch, still familiarizing myself with the codebase

I think your idea for ERC20 tokens is good